src/uct/ib/rc/accel/rc_mlx5_common.c

```diff
@@ -218,6 +218,181 @@ void uct_rc_mlx5_iface_common_tag_cleanup(uct_rc_mlx5_iface_common_t *iface,
 #endif
 }
 
+
+#if HAVE_IBV_EXP_DM
+const uint32_t uct_rc_mlx5_dm_key = 1245; /* magic number */
+
+typedef struct {
+    uct_mlx5_dm_data_t *dm;
+    int                count;
+} uct_iface_dm_mp_priv_t;
+
+typedef struct uct_mlx5_dm_va {
+    struct ibv_exp_dm  ibv_dm;
+    size_t             length;
+    uint64_t           *start_va;
+} uct_mlx5_dm_va_t;
+
+static ucs_status_t
+ucs_iface_dm_mpool_chunk_malloc(ucs_mpool_t *mp, size_t *size_p, void **chunk_p)
+{
+    ucs_status_t status;
+
+    status = ucs_mpool_chunk_malloc(mp, size_p, chunk_p);
+    if (status == UCS_OK) {
+        memset(*chunk_p, 0, *size_p);
+    }
+
+    return status;
+}
+
+static inline uct_iface_dm_mp_priv_t* uct_iface_dm_mp_priv(ucs_mpool_t *mp)
+{
+    return (uct_iface_dm_mp_priv_t*)ucs_mpool_priv(mp);
+}
+
+static void uct_iface_dm_mp_obj_init(ucs_mpool_t *mp, void *obj, void *chunk)
+{
+    uct_mlx5_dm_data_t *dm         = uct_iface_dm_mp_priv(mp)->dm;
+    uct_rc_iface_send_desc_t* desc = (uct_rc_iface_send_desc_t*)obj;
+
+    if ((desc->super.buffer == NULL) && (uct_iface_dm_mp_priv(mp)->count < dm->seg_count)) {
+        desc->lkey          = dm->mr->lkey;
+        desc->super.buffer  = UCS_PTR_BYTE_OFFSET(dm->start_va,
+                                                  uct_iface_dm_mp_priv(mp)->count * dm->seg_len);
+        desc->super.handler = (uct_rc_send_handler_t)ucs_mpool_put;
+        uct_iface_dm_mp_priv(mp)->count++;
+    }
+}
+
+static ucs_mpool_ops_t uct_dm_iface_mpool_ops = {
+    .chunk_alloc   = ucs_iface_dm_mpool_chunk_malloc,
+    .chunk_release = ucs_mpool_chunk_free,
+    .obj_init      = uct_iface_dm_mp_obj_init,
+    .obj_cleanup   = NULL
+};
+
+
+static int uct_mlx5_iface_dm_device_cmp(uct_mlx5_dm_data_t *dm_data,
+                                        uct_rc_iface_t *iface,
+                                        const uct_common_mlx5_iface_config_t *config)
+{
+    uct_ib_device_t *dev = uct_ib_iface_device(&iface->super);
+
+    return dm_data->device->ibv_context == dev->ibv_context;
+}
+
+static ucs_status_t uct_mlx5_iface_init_dm_tl(uct_mlx5_dm_data_t *data,
+                                              uct_rc_iface_t *iface,
```

Comment:
we could pass device here instead of iface

Reply:
we can't: we need iface->pd to register memory